### PR TITLE
fix: Allow Proxy Paths to have arbitrary names

### PR DESCRIPTION
**Pull request by jfuss**
*Issue #, if available:*
Addressed the bug from #457 where SAM CLI only accepted proxy paths with '{proxy+}'. This PR will allow 'proxy' to be any arbitrary name.

I ran the functional tests locally in both Py2 and Py3 (all passed in both).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/local/apigw/path_converter.py [hunk 1]
```diff
@@ -4,16 +4,31 @@
 
 import re
 
-FLASK_PATH_PARAMS = "/<path:proxy>"
-APIGW_PATH_PARAMS_ESCAPED = r"/{proxy\+}"
-APIGW_PATH_PARAMS = "/{proxy+}"
+# The regex captures any path information before the {proxy+}. This is to support paths that have other params in
+# them. Example: /id/{id}/user/{proxy+}. Otherwise the regex will match the first { with the last +} giving incorrect
+# results.
+APIGW_PATH_PARAMS_ESCAPED = r"(.*/){(.*)\+}"
```
Comment (sanathkr) [inline samcli/local/apigw/path_converter.py]: This is not any path param, but only wildcard ones ie. the ones with + at the end. Right? 

Can you update the comment and variable name to be more descriptive?

File: samcli/local/apigw/path_converter.py [hunk 2]
```diff
@@ -4,16 +4,31 @@
...
+
+# The regex replaces what was captured with APIGW_PATH_PARAMS_ESCAPED to construct the full path with the {proxy+}
+# replaces. The first group is anything before {proxy+}, while the second group is the name given to the proxy.
+# Example: /id/{id}/user/{resource+}; g<1> = '/id/{id}/user/'; g<2> = 'resource'
+FLASK_PATH_PARAMS = r"\g<1><path:\g<2>>"
```
Comment (sanathkr) [inline samcli/local/apigw/path_converter.py]: nit: `PATH_PARAMS` suffix isn't too clear. Consider dropping the suffix or changing to something else

File: tests/unit/local/apigw/test_path_converter.py [hunk 3]
```diff
@@ -19,6 +19,13 @@ def test_proxy_path(self):
 
         self.assertEquals(flask_path, "/<path:proxy>")
 
+    def test_proxy_path_with_different_name(self):
```
Comment (sanathkr) [inline tests/unit/local/apigw/test_path_converter.py]: Can you also test the following cases (mostly testing combination of single & multiple path params with wildcard

1. `/a/{id}/b/{resource+}`
2. `/a/b/{proxy}/{resource+}`
3. `/{id}/{something+}`
4. `/{a}/{b}/{c}/{d+}`

File: samcli/local/apigw/path_converter.py (hunk 1 above)
Comment (jfuss) [reply samcli/local/apigw/path_converter.py]: Updated name and added a full example.

The comment seems correct though. It is describing that it captures the path before the {proxy+}. Without this capture, the regex will match the first '{' and the last '+}', which is incorrect if there are path parameters in the full path. 

Is there a part of this that is confusing that could be cleared up further?

File: samcli/local/apigw/path_converter.py (hunk 2 above)
Comment (jfuss) [reply samcli/local/apigw/path_converter.py]: Updated to: `FLASK_CAPTURE_ALL_PATH`

File: tests/unit/local/apigw/test_path_converter.py (hunk 3 above)
Comment (jfuss) [reply tests/unit/local/apigw/test_path_converter.py]: Sure can do :)

**Comment by jfuss** [comment]
>Can you add more test cases? The regex is comprehensive (ie. uses (.*)), so it will be helpful to verify that the regex doesn't eat the whole string :)

test_proxy_with_path_param tested that it wouldn't eat the whole string but added the ones you suggested for better confidence. :)

File: samcli/local/apigw/path_converter.py (hunk 1 above)
Comment (sanathkr) [reply samcli/local/apigw/path_converter.py]: > path information before the {proxy+}
The "before" word gave me the impression that it is used to get everything in the string before the substring {proxy+}. Given /id/{id}/user/{proxy+}, is this regex used to return `id` and `proxy`? or just `proxy`?

File: samcli/local/apigw/path_converter.py [hunk 4]
```diff
@@ -4,16 +4,31 @@
 
 import re
 
-FLASK_PATH_PARAMS = "/<path:proxy>"
-APIGW_PATH_PARAMS_ESCAPED = r"/{proxy\+}"
-APIGW_PATH_PARAMS = "/{proxy+}"
+# The regex captures any path information before the {proxy+}. This is to support paths that have other params in
+# them. Otherwise the regex will match the first { with the last +} giving incorrect results.
+# Example: /id/{id}/user/{proxy+}. g<1> = '/id/{id}/user/' g<2> = 'proxy'
+PROXY_PATH_PARAMS_ESCAPED = r"(.*/){(.*)\+}"
```
Comment (sanathkr) [inline samcli/local/apigw/path_converter.py]: updated example looks better to me :)

File: tests/unit/local/apigw/test_path_converter.py [hunk 5]
```diff
@@ -19,6 +21,17 @@ def test_proxy_path(self):
 
         self.assertEquals(flask_path, "/<path:proxy>")
 
+    @parameterized.expand([("/{resource+}", "/<path:resource>"),
+                           ("/a/{id}/b/{resource+}", "/a/<id>/b/<path:resource>"),
+                           ("/a/b/{proxy}/{resource+}", "/a/b/<proxy>/<path:resource>"),
+                           ("/{id}/{something+}", "/<id>/<path:something>"),
+                           ("/{a}/{b}/{c}/{d+}", "/<a>/<b>/<c>/<path:d>")
```
Comment (sanathkr) [inline tests/unit/local/apigw/test_path_converter.py]: awesome!

**Pull request merged**